File: core/src/main/java/net/dontdrinkandroot/wicket/component/form/AbstractTemporalAccessorTextField.java
/*
 * Copyright (C) 2012-2017 Philip Washington Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.component.form;

import org.apache.wicket.markup.ComponentTag;
import org.apache.wicket.markup.html.form.TextField;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.Model;
import org.apache.wicket.util.convert.IConverter;
import org.apache.wicket.util.value.IValueMap;

import java.io.Serializable;
import java.time.temporal.TemporalAccessor;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class AbstractTemporalAccessorTextField<T extends TemporalAccessor & Serializable> extends TextField<T>
{
    private IModel<T> minModel = new Model<>();

    private IModel<T> maxModel = new Model<>();

    public AbstractTemporalAccessorTextField(String id, Class<T> type)
    {
        super(id, type);
    }

    public AbstractTemporalAccessorTextField(String id, IModel<T> model, Class<T> type)
    {
        super(id, model, type);
    }

    public void setMin(T min)
    {
        this.minModel.setObject(min);
    }

    public void setMax(T max)
    {
        this.maxModel.setObject(max);
    }

    public void setMinModel(IModel<T> minModel)
    {
        this.minModel = minModel;
    }

    public void setMaxModel(IModel<T> maxModel)
    {
        this.maxModel = maxModel;
    }

    @Override
    protected void onComponentTag(ComponentTag tag)
    {
        super.onComponentTag(tag);

        IConverter<T> converter = this.getConverter(this.getType());
        IValueMap attributes = tag.getAttributes();
        final T min = this.minModel.getObject();
        if (min != null) {
            attributes.put("min", converter.convertToString(min, this.getLocale()));
        } else {
            attributes.remove("min");
        }

        final T max = this.maxModel.getObject();
        if (max != null) {
            attributes.put("max", converter.convertToString(max, this.getLocale()));
        } else {
            attributes.remove("max");
        }
    }
}

File: core/src/main/kotlin/net/dontdrinkandroot/wicket/component/form/AbstractTemporalAccessorTextField.kt
package net.dontdrinkandroot.wicket.component.form

import org.apache.wicket.markup.ComponentTag
import org.apache.wicket.markup.html.form.TextField
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model
import java.io.Serializable
import java.time.temporal.TemporalAccessor

open class AbstractTemporalAccessorTextField<T>(
    id: String,
    model: IModel<T>? = null,
    type: Class<T>? = null,
    private val minModel: IModel<T> = Model(),
    private val maxModel: IModel<T> = Model()
) :
    TextField<T>(id, model, type) where T : TemporalAccessor, T : Serializable {

    fun setMin(min: T) {
        minModel.setObject(min)
    }

    fun setMax(max: T) {
        maxModel.setObject(max)
    }

    override fun onComponentTag(tag: ComponentTag) {
        super.onComponentTag(tag)
        val converter = getConverter(this.type)
        val attributes = tag.attributes
        val min = minModel.getObject()
        if (min != null) {
            attributes["min"] = converter.convertToString(min, this.locale)
        } else {
            attributes.remove("min")
        }
        val max = maxModel.getObject()
        if (max != null) {
            attributes["max"] = converter.convertToString(max, this.locale)
        } else {
            attributes.remove("max")
        }
    }
}